packages/core/src/browser/shell/application-shell.ts

```diff
@@ -914,6 +916,28 @@ export class ApplicationShell extends Widget {
             }
             // Set the z-index so elements with `position: fixed` contained in the active widget are displayed correctly
             this.setZIndex(newValue.node, '1');
+
+            // activate another widget if an active widget gets closed
+            const onCloseRequest = newValue['onCloseRequest'];
+            newValue['onCloseRequest'] = msg => {
+                const currentTabBar = this.currentTabBar;
+                if (currentTabBar) {
+                    const recentlyUsedInTabBar = currentTabBar['_previousTitle'] as TabBar<Widget>['currentTitle'];
+                    if (recentlyUsedInTabBar && recentlyUsedInTabBar.owner !== newValue) {
+                        currentTabBar.currentIndex = ArrayExt.firstIndexOf(currentTabBar.titles, recentlyUsedInTabBar);
+                        if (currentTabBar.currentTitle) {
+                            currentTabBar.currentTitle.owner.activate();
+                        }
+                    } else if (!this.activateNextTabInTabBar(currentTabBar)) {
+                        if (!this.activatePreviousTabBar(currentTabBar)) {
+                            this.activateNextTabBar(currentTabBar);
+                        }
+                    }
+                }
+                newValue['onCloseRequest'] = onCloseRequest;
+                newValue['onCloseRequest'](msg);
+            };
+            this.toDisposeOnActiveChanged.push(Disposable.create(() => newValue['onCloseRequest'] = onCloseRequest));
```

Comment:
In case we fork or reimplement PhosphorJS, we should add direct support for this so we don't have to fiddle with the `onCloseRequest` method.

Reply:
Yeah something like `onWillClose` event for all widgets will be better.